### PR TITLE
Adding support for Remote .NET Attach and tsc fix on Windows

### PR DESCRIPTION
**Pull request by WardenGnaw**
**Feature:**
package.json adds listRemoteProcesses as a command, including debuggerPath into pipeTransport. The
command is registered into commands.ts. processPicker.ts has its process listing parser refactored
and now includes calling process listing to remote processes. Only works on Linux and OSX. Windows
remote attach will result in a rejected promise.

**Fix:**
Fixing tsc error TS1252 in platform.ts.
Moved the getValue function within getCurrentPlatform function outside of function scope. Added string array parameter to getValue function because lines was originally a local variable within the getCurrentPlatform function.

### CODE REVIEW

File: src/features/processPicker.ts [hunk 1]
```diff
@@ -35,6 +37,80 @@ export class AttachPicker {
     }
 }
 
+export class RemoteAttachPicker {
+    public static ShowAttachEntries(args : any): Promise<string> {
+        // Grab selected name from UI
+        var name : string = args.name;
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: Do we need to error out if `name` is undefined, not a string, or the empty string?

File: src/features/processPicker.ts [hunk 2]
```diff
@@ -35,6 +37,80 @@ export class AttachPicker {
...
+        // Find correct pipeTransport via selected name
+        var config; 
+        for (var i = 0; i < json.configurations.length; ++i) {
+            if (json.configurations[i].name === name) {
+                config = json.configurations[i];
+                break; 
+            }
+        }
+
+        if (!config.pipeTransport) {
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: Error if we couldn't find the right config?

File: src/features/processPicker.ts [hunk 3]
```diff
@@ -35,6 +37,80 @@ export class AttachPicker {
...
+            // Missing PipeTransport, prompt if user wanted to just do local attach?
+            return new Promise<string>((resolve, reject) => {
+                reject(new Error("Configuration \"" + args.name + "\" in launch.json does not have a " + 
+                "pipeTransport argument for pickRemoteProcess. Use pickProcess for local attach."));
+            });
+        } else {
+            var pipeCmd : string = config.pipeTransport.pipeProgram + " " + config.pipeTransport.pipeArgs[0];
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: You want to join the arguments together, not just take the 0th entry from the array.

File: src/features/processPicker.ts [hunk 4]
```diff
@@ -35,6 +37,80 @@ export class AttachPicker {
...
+            // Clean string of newlines
+            var cleanOutput : string = output.replace(/[\r\n]+/g, '').toLowerCase();
+
+            switch(cleanOutput) {
+                case "darwin":
+                case "linux":
+                    return cleanOutput;
+                // Failure case. TODO: test for windows machine
+                default:
+                    return new Promise<string>((resolve, reject) => {
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: You are going to need an error string here.

File: src/features/processPicker.ts [hunk 5]
```diff
@@ -64,28 +140,47 @@ abstract class DotNetAttachItemsProvider implements AttachItemsProvider {
...
+
 export class PsAttachItemsProvider extends DotNetAttachItemsProvider {
+    protected getInternalProcessEntries(): Promise<Process[]> {
+        const commColumnTitle = Array(PsOutputParser.secondColumnCharacters).join("a");
+        // the BSD version of ps uses '-c' to have 'comm' only output the executable name and not
+        // the full path. The Linux version of ps has 'comm' to only display the name of the executable
+        // Note that comm on Linux systems is truncated to 16 characters:
+        // https://bugzilla.redhat.com/show_bug.cgi?id=429565
+        // Since 'args' contains the full path to the executable, even if truncated, searching will work as desired.
+        const psCommand = `ps -axww -o pid=,comm=${commColumnTitle},args=` + (os.platform() === 'darwin' ? ' -c' : '');
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: as.platform() is going to tell you the host platform, you want the target platform

File: src/features/processPicker.ts [hunk 6]
```diff
@@ -35,6 +37,80 @@ export class AttachPicker {
...
+
+        if (!config.pipeTransport) {
+            // Missing PipeTransport, prompt if user wanted to just do local attach?
+            return new Promise<string>((resolve, reject) => {
+                reject(new Error("Configuration \"" + args.name + "\" in launch.json does not have a " + 
+                "pipeTransport argument for pickRemoteProcess. Use pickProcess for local attach."));
+            });
+        } else {
+            var pipeCmd : string = config.pipeTransport.pipeProgram + " " + config.pipeTransport.pipeArgs[0];
+            return RemoteAttachPicker.getRemoteOS(pipeCmd).then(remoteOS => {
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: At least for some pipes (ex: ssh) you are going to get way better perf if you can execute the pipe once and have it return both the OS arch and the PS output. This might turn out to be too complicated, but it is probably worth a try. One thing you could do is optimize for Linux and just require multiple round trips on OSX. Something like:

uname && [ $(uname) == "Linux" ] && ps ...

File: src/features/processPicker.ts [hunk 7]
```diff
@@ -246,4 +333,16 @@ function execChildProcess(process: string, workingDirectory: string): Promise<st
             resolve(stdout);
         });
     });
+}
+
+function execChildProcessWithTimeout(process: string, workingDirectory: string): Promise<string> {
+    return new Promise<string>((resolve, reject) => {
+        child_process.exec(process, { timeout: 3000, cwd: workingDirectory, maxBuffer: 500 * 1024 }, (error: Error, stdout: string, stderr: string) => {
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: 3 seconds sounds much too small

Comment (gregg-miskelly) [inline src/features/processPicker.ts]: Is 'maxBuffer' the maximum number of characters to hold? If so, do we need to specify a max length?

File: src/features/processPicker.ts [hunk 8]
```diff
@@ -246,4 +333,16 @@ function execChildProcess(process: string, workingDirectory: string): Promise<st
...
+            if (stderr && stderr.length > 0) {
+                reject(stderr);
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: We may well need to prepend this with some text for this to be comprehensible. Something like 'Pipe transport program bla failed.\n\n<message>'

File: src/platform.ts [hunk 9]
```diff
@@ -34,33 +56,12 @@ export function getCurrentPlatform() {
         const text = child_process.execSync('cat /etc/os-release').toString();
         const lines = text.split('\n');
 
-        function getValue(name: string) {
```
Comment (gregg-miskelly) [inline src/platform.ts]: Is this change because TypeScript wasn't working for you? This should probably go in as a separate PR.

File: src/features/processPicker.ts (hunk 5 above)
Comment (WardenGnaw) [inline src/features/processPicker.ts]: processPicker.ts has the code for both local and remote attach listing. getInternalProcessEntries is used for local attach.

File: src/features/processPicker.ts (hunk 7 above)
Comment (WardenGnaw) [inline src/features/processPicker.ts]: What would be the recommended timeout for waiting on a remote command?

**Comment by chuckries** [comment]
Please run 'gulp tslint' and fix any linting issues in the files you are touching. Many of your 'var's should be 'let's.

File: src/features/processPicker.ts (hunk 7 above)
Comment (WardenGnaw) [inline src/features/processPicker.ts]: I am not sure. I used execChildProcess as an example of calling exec on a child process.

Comment (gregg-miskelly) [inline src/features/processPicker.ts]: Will this command be async? Is there a way to cancel it? Its always nice to not have timeouts at all, as it is always hard to pick a good one. If you need one, I would probably suggest picking something pretty long since this involves connecting to a remote machine, which could be fairly slow. Maybe 30 seconds?

**Comment by rajkumar42** [comment]
@edmunoz is added to the review. #Closed

File: package.json [hunk 10]
```diff
@@ -638,6 +669,178 @@
...
+                    "default": {}
+                  },
+                  "windows": {
+                    "type": "object",
+                    "description": "Windows-specific pipe launch configuration options",
+                    "default": {
+                      "pipeCwd": "${workspaceRoot}",
+                      "pipeProgram": "enter the fully qualified path for the pipe program name, for example 'c:\\tools\\plink.exe'",
+                      "pipeArgs": [],
+                      "debuggerPath" : "enter the path for the debugger, for example ~/clrdbg/clrdbg"
```
Comment (gregg-miskelly) [inline package.json]: Thinking about this some more, there is no reason why 'debuggerPath' would be dependent on the host OS. So I would suggest removing it as a property from within the 'windows', 'osx', and 'linux' elements.

File: package.json [hunk 11]
```diff
@@ -638,6 +669,178 @@
                     "default": false
                   }
                 }
+              },
+              "pipeTransport": {
+                "type": "object",
```
Comment (gregg-miskelly) [inline package.json]: In the attach case, I think we should make 'debuggerPath' a required property.

File: package.json [hunk 12]
```diff
@@ -440,7 +453,8 @@
                     "default": {
                       "pipeCwd": "${workspaceRoot}",
                       "pipeProgram": "enter the fully qualified path for the pipe program name, for example 'c:\\tools\\plink.exe'",
-                      "pipeArgs": []
+                      "pipeArgs": [],
+                      "debuggerPath" : "Please enter the debugger path here, for example c:\\clrdbg\\clrdbg"
```
Comment (gregg-miskelly) [inline package.json]: Same suggestion as in attach: I think we don't want the OS-specific 'debuggerPath' properties.

File: package.json [hunk 13]
```diff
@@ -426,6 +434,11 @@
                     },
                     "default": []
                   },
+                  "debuggerPath" : {
+                    "type" : "string",
+                    "description" : "The full path to the debugger",
```
Comment (gregg-miskelly) [inline package.json]: I would suggest adding to the end of this description 'on the target computer' (or similar if you don't like that description)

File: src/features/processPicker.ts [hunk 14]
```diff
@@ -35,6 +37,111 @@ export class AttachPicker {
...
+            });
+        }
+
+        if (!config.pipeTransport) {
+            // Missing PipeTransport, prompt if user wanted to just do local attach.
+            return new Promise<string>((resolve, reject) => {
+                reject(new Error("Configuration \"" + args.name + "\" in launch.json does not have a " + 
+                "pipeTransport argument for pickRemoteProcess. Use pickProcess for local attach."));
+            });
+        } else {
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: I think I would suggest verifying that the 'debuggerPath' property is present. Since if the user has accidentally included the command line to execute the debugger in the pipeArgs, this is going to do pretty confusing things.

File: src/features/processPicker.ts [hunk 15]
```diff
@@ -35,6 +37,111 @@ export class AttachPicker {
...
+            let pipeCmd : string = config.pipeTransport.pipeProgram + " " + config.pipeTransport.pipeArgs.join(" ");
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: Both pipeArgs and pipeProgram can be overwritten with the host-OS-specific options. So you need to check if those properties exist, and if so, give preference to them. See OpenDebugAD7 for an example.

File: src/features/processPicker.ts [hunk 16]
```diff
@@ -35,6 +37,111 @@ export class AttachPicker {
...
+                return new Promise<AttachItem[]>((resolve, reject) => {
+                    reject(new Error("Pipe transport failed to get OS and processes."));
+                });
+            }
+            else {
+                let remoteOS = lines[0].replace(/[\r\n]+/g, '').toLowerCase();
+                
+                if (remoteOS != "linux" && remoteOS != "darwin") {
+                    return new Promise<AttachItem[]>((resolve, reject) => {
+                        reject(new Error("Pipe Transport attach could not determine os. " + remoteOS));
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: Or the operating system is not supported.

File: src/features/processPicker.ts [hunk 17]
```diff
@@ -35,6 +37,111 @@ export class AttachPicker {
...
+            // Processess will follow if listed
+            let lines = output.split(os.EOL);
+            
+            if (lines.length == 0) {
+                return new Promise<AttachItem[]>((resolve, reject) => {
+                    reject(new Error("Pipe transport failed to get OS and processes."));
+                });
+            }
+            else {
+                let remoteOS = lines[0].replace(/[\r\n]+/g, '').toLowerCase();
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: Might want to NOT convert this to lower case since the bash script doesn't.

File: src/features/processPicker.ts [hunk 18]
```diff
@@ -35,6 +37,111 @@ export class AttachPicker {
...
+                
+                if (remoteOS != "linux" && remoteOS != "darwin") {
+                    return new Promise<AttachItem[]>((resolve, reject) => {
+                        reject(new Error("Pipe Transport attach could not determine os. " + remoteOS));
+                    }); 
+                }
+                
+                // Only got OS from uname
+                if (lines.length == 1) {
+                    return RemoteAttachPicker.getRemoteProcesses(pipeCmd, remoteOS);
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: Shouldn't this really just be an error?

File: package.json (hunk 11 above)
Comment (gregg-miskelly) [inline package.json]: Attach with pipeTransport. There is a 'required' property that you should be able to add here to do that.

File: package.json [hunk 19]
```diff
@@ -153,6 +154,11 @@
         "command": "csharp.listProcess",
         "title": "List process for attach",
         "category": "CSharp"
+      },
+      {
+        "command" : "csharp.listRemoteProcess",
+        "title" : "List processes on remote connection for attach",
+        "category": "CSharp"
```
Comment (DustinCampbell) [inline package.json]: I notice that this command and "csharp.listProcess" have a category of "CSharp", but chsarp.downloadDebugger" has a category of "Debug". Is that intentional? Also, should the category be "C#"?

File: src/features/commands.ts [hunk 20]
```diff
@@ -36,8 +36,9 @@ export default function registerCommands(server: OmnisharpServer, extensionPath:
     let attachItemsProvider = DotNetAttachItemsProviderFactory.Get();
     let attacher = new AttachPicker(attachItemsProvider);
     let d8 = vscode.commands.registerCommand('csharp.listProcess', () => attacher.ShowAttachEntries());
+    let d9 = vscode.commands.registerCommand('csharp.listRemoteProcess', (args) => RemoteAttachPicker.ShowAttachEntries(args));
 
-    return vscode.Disposable.from(d1, d2, d3, d4, d5, d6, d7, d8);
+    return vscode.Disposable.from(d1, d2, d3, d4, d5, d6, d7, d8, d9);
```
Comment (DustinCampbell) [inline src/features/commands.ts]: This will conflict with #795. First one merged wins! :smile:

File: src/features/processPicker.ts [hunk 21]
```diff
@@ -35,6 +37,136 @@ export class AttachPicker {
...
+export class RemoteAttachPicker {
+    public static ShowAttachEntries(args : any): Promise<string> {
+        // Grab selected name from UI
+        let name : string = args.name;
+
+        if (!name) {
+            // Config name not found. 
+            return new Promise<string>((resolve, reject) => {
+                reject(new Error("Name not defined in current configuration."));
+            });
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]: 1. This could be written simpler as

```JS
return Promise.reject<string>(new Error("Name not defined in current configuration"));
```

Comment (DustinCampbell) [inline src/features/processPicker.ts]: 2. Should the error message mention the name that was passed in?

File: src/features/processPicker.ts [hunk 22]
```diff
@@ -35,6 +37,136 @@ export class AttachPicker {
...
+            if (json.configurations[configIdx].name === name) {
+                config = json.configurations[configIdx];
+                break; 
+            }
+        }
+
+        if (configIdx == json.configurations.length) {
+            // Name not found in list of given configurations. 
+            return new Promise<string>((resolve, reject) => {
+                reject(new Error("Could not find configuration that matches given name"));
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]: Similar feedback to my comments above.

File: src/features/processPicker.ts [hunk 23]
```diff
@@ -35,6 +37,136 @@ export class AttachPicker {
...
+            });
+        }
+
+        if (!config.pipeTransport || !config.pipeTransport.debuggerPath) {
+            // Missing PipeTransport and debuggerPath, prompt if user wanted to just do local attach.
+            return new Promise<string>((resolve, reject) => {
+                reject(new Error("Configuration \"" + args.name + "\" in launch.json does not have a " + 
+                "pipeTransport argument with debuggerPath for pickRemoteProcess. Use pickProcess for local attach."));
+            });
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]: Same here too.

File: src/features/processPicker.ts [hunk 24]
```diff
@@ -35,6 +37,136 @@ export class AttachPicker {
...
+        } else if (osPlatform == "linux" && config.pipeTransport.linux) {
+            return config.pipeTransport.linux;
+        } else if (osPlatform == "win32" && config.pipeTransport.windows) {
+            return config.pipeTransport.windows;
+        }
+        
+        return null;
+    }
+    
+    public static getRemoteOSAndProcesses(pipeCmd :string) : Promise<AttachItem[]> {
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]: spacing looks wrong here. Might be worth a <kbd>Shift+Alt+F</kbd> in VS Code.

File: src/features/processPicker.ts [hunk 25]
```diff
@@ -35,6 +37,136 @@ export class AttachPicker {
...
+        const commColumnTitle = Array(PsOutputParser.secondColumnCharacters).join("a");
+        // Commands to get OS and processes
+        const linuxPsCommand = `ps -axww -o pid=,comm=${commColumnTitle},args=`;
+        const osxPsCommand = `ps -axww -o pid=,comm=${commColumnTitle},args= -c`;
+        const command = `uname && if [ $(uname) == "Linux" ] ; then ` + linuxPsCommand + `; elif [ $(uname) == "Darwin" ] ; ` + 
+         `then ` + osxPsCommand + `; fi`;
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]: Why use `` string literals, but then not use placeholders?

```TypeScript
const command = `uname && if [ $(uname) == "Linux" ] ; then ${linuxPsCommand}; elif [ $(uname) == "Darwin" ] ; then ${osxPsCommand}; fi`;
```

File: src/features/processPicker.ts [hunk 26]
```diff
@@ -35,6 +37,136 @@ export class AttachPicker {
...
+
+        return execChildProcess(pipeCmd + ' "' + command + '"', null).then(output => {
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]: Good spot for a `` string literal.

File: src/features/processPicker.ts [hunk 27]
```diff
@@ -35,6 +37,136 @@ export class AttachPicker {
...
+                    reject(new Error("Pipe transport failed to get OS and processes."));
+                });
+            }
+            else {
+                let remoteOS = lines[0].replace(/[\r\n]+/g, '');
+                
+                if (remoteOS != "Linux" && remoteOS != "Darwin") {
+                    return new Promise<AttachItem[]>((resolve, reject) => {
+                        reject(new Error("Could not determine operating system or operating system not supported. " + remoteOS));
+                    });
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]:
```TypeScript
return Promise.reject<AttackItem[]>(new Error(`Could not determine operating system or operating system not supported. ${remoteOS}`));
```

File: src/features/processPicker.ts [hunk 28]
```diff
@@ -35,6 +37,136 @@ export class AttachPicker {
...
+                    return new Promise<AttachItem[]>((resolve, reject) => {
+                        reject(new Error("Could not determine operating system or operating system not supported. " + remoteOS));
+                    }); 
+                }
+                
+                // Only got OS from uname
+                if (lines.length == 1) {
+                    return new Promise<AttachItem[]>((resolve, reject) => {
+                        reject(new Error("Transport attach could not obtain processes list.")); 
+                    });
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]: Same as above. You can just return `Promise.reject`.

File: src/features/processPicker.ts [hunk 29]
```diff
@@ -35,6 +37,136 @@ export class AttachPicker {
...
+                    let processes = lines.slice(1);
+                    return sortProcessEntries(PsOutputParser.parseProcessFromPsArray(processes), remoteOS);
+                }
+            }
+        });
+    }
+
+    public static getRemoteProcesses(pipeCmd: string, os: string) : Promise<AttachItem[]> {
+        const commColumnTitle = Array(PsOutputParser.secondColumnCharacters).join("a");
+        const psCommand = `ps -axww -o pid=,comm=${commColumnTitle},args=` + (os === 'darwin' ? ' -c' : '');
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]: Also could be:

```TypeScript
const psCommand = `ps -axww -o pid=,comm=${commColumnTitle},args=${(os === 'darwin' ? ' -c' : '')}`;
```

File: src/features/processPicker.ts [hunk 30]
```diff
@@ -35,6 +37,136 @@ export class AttachPicker {
...
+
+        return execChildProcess(pipeCmd + ' "' + command + '"', null).then(output => {
+            // OS will be on first line
+            // Processess will follow if listed
+            let lines = output.split(os.EOL);
+            
+            if (lines.length == 0) {
+                return new Promise<AttachItem[]>((resolve, reject) => {
+                    reject(new Error("Pipe transport failed to get OS and processes."));
+                });
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]: Another spot to return `Promise.reject`.

File: src/features/processPicker.ts [hunk 31]
```diff
@@ -64,28 +196,47 @@ abstract class DotNetAttachItemsProvider implements AttachItemsProvider {
...
+
 export class PsAttachItemsProvider extends DotNetAttachItemsProvider {
+    protected getInternalProcessEntries(): Promise<Process[]> {
+        const commColumnTitle = Array(PsOutputParser.secondColumnCharacters).join("a");
+        // the BSD version of ps uses '-c' to have 'comm' only output the executable name and not
+        // the full path. The Linux version of ps has 'comm' to only display the name of the executable
+        // Note that comm on Linux systems is truncated to 16 characters:
+        // https://bugzilla.redhat.com/show_bug.cgi?id=429565
+        // Since 'args' contains the full path to the executable, even if truncated, searching will work as desired.
+        const psCommand = `ps -axww -o pid=,comm=${commColumnTitle},args=` + (os.platform() === 'darwin' ? ' -c' : '');
```
Comment (DustinCampbell) [inline src/features/processPicker.ts]: Is this code copied and pasted? Could it be shared?

File: src/features/processPicker.ts (hunk 21 above)
Comment (WardenGnaw) [inline src/features/processPicker.ts]: This is the case where name is null.

File: package.json (hunk 19 above)
Comment (WardenGnaw) [inline package.json]: This I am not to sure of, @rajkumar42 or @gregg-miskelly

File: src/features/processPicker.ts (hunk 21 above)
Comment (DustinCampbell) [inline src/features/processPicker.ts]: Ah of course. I was reading quickly. :smile:

File: package.json (hunk 19 above)
Comment (gregg-miskelly) [inline package.json]: I don't know what `category` is supposed to be for. You can see if the schema will tell you. @chuckries created the downladDebugger command, so he might know. Otherwise you can look at the VS Code source to see if that has the answer (http://github.com/Microsoft/VSCode)

File: src/features/processPicker.ts [hunk 32]
```diff
@@ -35,6 +37,125 @@ export class AttachPicker {
...
+
+        return null;
+    }
+
+    public static getRemoteOSAndProcesses(pipeCmd: string): Promise<AttachItem[]> {
+        // Commands to get OS and processes
+        const command = `uname && if [ $(uname) == "Linux" ] ; then ${RemoteAttachPicker.linuxPsCommand} ; elif [ $(uname) == "Darwin" ] ; ` +
+            `then ${RemoteAttachPicker.osxPsCommand}; fi`;
+
+        return execChildProcess(`${pipeCmd} "${command}"`, null).then(output => {
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: I added this comment earlier, but to a spot where it got lost -- currently we will just reject the promise if the command returns stderror text. This might be confusing now that we are using execChildProcess for the pipe program. I would suggest doing something like putting in the wrong ip address in your `plink` or `ssh` command line and see if the experience is compressible.

File: src/features/processPicker.ts [hunk 33]
```diff
@@ -35,6 +37,125 @@ export class AttachPicker {
...
+            let lines = output.split(os.EOL);
+
+            if (lines.length == 0) {
+                return Promise.reject<AttachItem[]>(new Error("Pipe transport failed to get OS and processes."));
+            }
+            else {
+                let remoteOS = lines[0].replace(/[\r\n]+/g, '');
+
+                if (remoteOS != "Linux" && remoteOS != "Darwin") {
+                    return Promise.reject<AttachItem[]>(new Error("Could not determine operating system " +
```
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: Doesn't this always been the OS is unsupported?

File: src/platform.ts (hunk 9 above)
Comment (WardenGnaw) [inline src/platform.ts]: Done in #801

File: package.json (hunk 19 above)
Comment (DustinCampbell) [inline package.json]: `category` appears in the Command Palette:

image

File: src/features/processPicker.ts [hunk 34]
```diff
@@ -246,4 +418,27 @@ function execChildProcess(process: string, workingDirectory: string): Promise<st
...
+        child_process.exec(process, { cwd: workingDirectory, maxBuffer: 500 * 1024 }, (error: Error, stdout: string, stderr: string) => {
+            if (error) {
+                reject(new Error("See remote-attach output."));
+                channel.append(error.message);
+                channel.show();
+                return;
+            }
+
+            if (stderr && stderr.length > 0) {
+                reject(new Error("See remote-attach output"));
```
Comment (WardenGnaw) [inline src/features/processPicker.ts]: I'm confused about how to do this. I thought this would be part of the callback function after it finishes the exec.

File: src/features/commands.ts (hunk 20 above)
Comment (WardenGnaw) [inline src/features/commands.ts]: I lost. 😢

File: src/features/processPicker.ts [hunk 35]
```diff
@@ -246,4 +418,27 @@ function execChildProcess(process: string, workingDirectory: string): Promise<st
...
+    return new Promise<string>((resolve, reject) => {
+        child_process.exec(process, { cwd: workingDirectory, maxBuffer: 500 * 1024 }, (error: Error, stdout: string, stderr: string) => {
+            if (error) {
+                reject(new Error("See remote-attach output."));
+                channel.append(error.message);
+                channel.show();
+                return;
+            }
+
+            if (stderr && stderr.length > 0) {
```
Comment (WardenGnaw) [inline src/features/processPicker.ts]: Something like this? Using the channel as the buffer.
```
            if (stdout && stdout.length > 0) {
                channel.appendLine('stdout:');
                channel.append(stdout);
            }

            if (stderr && stderr.length > 0) {
                channel.appendLine('stderr:');
                channel.append(stderr);
            }

            if (error) {
                channel.appendLine('Error message:');
                channel.append(error.message);
            }


            if (error || (stderr && stderr.length > 0)) {
                reject(new Error("See remote-attach output"));
                channel.show();
                return;
            }

            resolve(stdout);
```

Comment (gregg-miskelly) [inline src/features/processPicker.ts]: Yes, except that you probably don't want to send it to the channel in the success case. So I would suggest keeping an internal string buffer until there is an error. You probably don't need the stderr/stdout prefix -- if something is an error or normal output is usually apparent to humans by reading the text.

File: src/features/processPicker.ts (hunk 34 above)
Comment (gregg-miskelly) [inline src/features/processPicker.ts]: I think you just keep state as to if you saw an error and then you can reject the promise at the end if there was an error.

File: package.json (hunk 19 above)
Comment (gregg-miskelly) [inline package.json]: Looking at https://github.com/Microsoft/vscode-node-debug/search?utf8=%E2%9C%93&q=pickNodeProcess, I wonder if there is a way to avoid declaring this command entirely since it shouldn't actually show up in the command well.

**Comment by gregg-miskelly** [comment]
This PR has been replaced with https://github.com/OmniSharp/omnisharp-vscode/pull/918

**Comment by asyschikov** [comment]
@WardenGnaw Is there any documentation on remote debugging, including how to use pipeTransport?

**Comment by gregg-miskelly** [comment]
@asyschikov I wrote up https://github.com/OmniSharp/omnisharp-vscode/wiki/Attaching-to-remote-processes. Let me know if you need more information/find problems.

**Comment by asyschikov** [comment]
@gregg-miskelly looks good thanks. Is debugging (locally) over SSH advantageous in any way?

**Comment by gregg-miskelly** [comment]
@asyschikov no - I wouldn't expect it to make a big difference, but it would add a bit of extra latency in every interaction between the UI and the backend.
